src/control/mobile_manipulation/mobile_manipulation_learner.py

```diff
@@ -0,0 +1,247 @@
+        :param val_env:  gym.Env, optional, if specified periodically evaluate on this env
+        :param logging_path: str, path for logging and checkpointing
+        :param silent: bool, disable verbosity
+        :param verbose: bool, enable verbosity
+        :return:
+        """
+        if logging_path == '':
+            logging_path = self.temp_path
+
+        if env is not None:
+            assert env.action_space == self.stable_bl_agent.env.action_space
+            assert env.observation_space == self.stable_bl_agent.env.observation_space
+            self.stable_bl_agent.env = env
+
+        rospy.loginfo("Start learning loop")
+        if val_env is not None:
+            eval_callback = MobileRLEvalCallback(eval_env=val_env,
+                                                 n_eval_episodes=self.nr_evaluations,
+                                                 eval_freq=self.evaluation_frequency,
+                                                 log_path=logging_path,
+                                                 best_model_save_path=logging_path,
+                                                 checkpoint_after_iter=self.checkpoint_after_iter,
+                                                 verbose=verbose if not silent else False)
+        else:
+            eval_callback = None
+        self.stable_bl_agent.learn(total_timesteps=self.iters,
+                                   callback=eval_callback,
+                                   eval_env=None)
+
+        self.stable_bl_agent.save(os.path.join(logging_path, f'last_model'))
+
+        for e in [env, val_env]:
+            if e is not None:
+                env.env_method("clear")
+        rospy.loginfo("Training finished")
+
+    def eval(self, env, name_prefix='', nr_evaluations: int = None):
+        """
+        Evaluate the agent on the specified environment.
+
+        :param env: gym.Env, env to evaluate on
+        :param name_prefix: str, name prefix for all logged variables
+        :param nr_evaluations: int, number of episodes to evaluate over
+        :return:
+        """
+        if nr_evaluations is None:
+            nr_evaluations = self.nr_evaluations
+        if isinstance(env, VecEnv):
+            assert env.num_envs == 1, "You must pass only one environment when using this function"
+            env = env.envs[0]
+
+        prefix = ''
+        episode_rewards, episode_lengths, metrics, name_prefix = evaluation_rollout(self.stable_bl_agent, env,
+                                                                                    nr_evaluations, name_prefix=prefix,
+                                                                                    global_step=self.stable_bl_agent.num_timesteps,
+                                                                                    verbose=2)
+        env.clear()
+        return episode_rewards, episode_lengths, metrics, name_prefix
```

Comment:
Please return a dictionary with these values, as described in the specifications.

Reply:
Done